File: scripts/frontmatter.py
#!/usr/bin/env python3
"""Closed parser for the Atlas §20.4 YAML-shaped grammar.

The public entry points accept bytes and either return one complete parsed
mapping or raise :class:`FrontmatterError`.  No decoding or newline
normalization is delegated to text-mode I/O.
"""
from __future__ import annotations

import json
import re
from dataclasses import dataclass
from pathlib import Path

MAX_DOCUMENT_BYTES = 131_072
MAX_FILE_BYTES = 262_144
MAX_LINE_BYTES = 4_096
MAX_SCALAR_BYTES = 8_192
MAX_DEPTH = 8
MAX_FIELDS = 64
MAX_SEQUENCE_ENTRIES = 1_024
MAX_NODES = 16_384

_KEY_VALUE_RE = re.compile(r"^([A-Za-z_][A-Za-z0-9_-]*):(?: (.*))?$")
_UNSUPPORTED_PREFIXES = ("&", "*", "!", "%")


class FrontmatterError(ValueError):
    """One deterministic, fail-closed §20.4 diagnostic."""


@dataclass(frozen=True)
class _Line:
    text: str
    number: int
    indent: int
    content: str

    @property
    def ignored(self) -> bool:
        return not self.content or self.content.startswith("#")


def _error(source: str | Path, line: int, message: str) -> FrontmatterError:
    return FrontmatterError(f"{source}: frontmatter line {line}: {message}")


def _line_of(data: bytes, offset: int) -> int:
    return data.count(b"\n", 0, offset) + 1


def _validate_file_bytes(data: bytes, source: str | Path) -> None:
    # Whole-file law is §25.8's (UTF-8, no BOM, LF only, the file ceiling);
    # tab/NUL/C0 are §20.4 frontmatter-region rules — the markdown body below
    # the closing fence is outside the grammar and may carry tabs.
    if len(data) > MAX_FILE_BYTES:
        raise _error(source, 1, f"whole file exceeds {MAX_FILE_BYTES} bytes")
    if data.startswith(b"\xef\xbb\xbf"):
        raise _error(source, 1, "UTF-8 BOM is unsupported")
    cr = data.find(b"\r")
    if cr >= 0:
        raise _error(source, _line_of(data, cr), "CR/CRLF is unsupported; use LF")


def _validate_utf8(data: bytes, source: str | Path) -> None:
    try:
        data.decode("utf-8", errors="strict")
    except UnicodeDecodeError as exc:
        raise _error(source, _line_of(data, exc.start), "input is not strict UTF-8") from None


def _fenced_region(data: bytes, source: str | Path) -> tuple[bytes, int]:
    """Return body bytes and its physical first-line number.

    The closing fence is located with bounded byte scanning before the input is
    decoded or split, so the document ceiling cannot be hidden by text I/O.
    """
    if not data.startswith(b"---\n"):
        raise _error(source, 1, "opening fence must be the exact line '---'")
    pos = 4
    while pos <= len(data):
        end = data.find(b"\n", pos)
        if end < 0:
            end = len(data)
        line = data[pos:end]
        if line == b"---":
            closing_end = end + (1 if end < len(data) else 0)
            if closing_end > MAX_DOCUMENT_BYTES:
                raise _error(
                    source,
                    _line_of(data, pos),
                    f"frontmatter exceeds {MAX_DOCUMENT_BYTES} bytes",
                )
            return data[4:pos], 2
        if end >= MAX_DOCUMENT_BYTES:
            raise _error(
                source,
                _line_of(data, pos),
                f"frontmatter exceeds {MAX_DOCUMENT_BYTES} bytes",
            )
        if end == len(data):
            break
        pos = end + 1
    raise _error(source, _line_of(data, len(data)), "missing closing fence '---'")


def _validate_region_bytes(
    region: bytes, source: str | Path, first_line: int
) -> list[_Line]:
    if len(region) > MAX_DOCUMENT_BYTES:
        raise _error(
            source, first_line, f"frontmatter exceeds {MAX_DOCUMENT_BYTES} bytes"
        )
    raw_lines = region.split(b"\n")
    lines: list[_Line] = []
    for offset, raw in enumerate(raw_lines):
        number = first_line + offset
        if len(raw) > MAX_LINE_BYTES:
            raise _error(source, number, f"line exceeds {MAX_LINE_BYTES} bytes")
        for index, byte in enumerate(raw):
            if byte == 0x09:
                raise _error(source, number, "tab is unsupported")
            if byte == 0x00:
                raise _error(source, number, "NUL is unsupported")
            if byte < 0x20:
                raise _error(source, number, f"C0 control 0x{byte:02x} is unsupported")
        try:
            text = raw.decode("utf-8", errors="strict")
        except UnicodeDecodeError:
            raise _error(source, number, "input is not strict UTF-8") from None
        indent = len(text) - len(text.lstrip(" "))
        if text[:indent] != " " * indent:
            raise _error(source, number, "indentation must use ASCII spaces")
        if indent % 2:
            raise _error(source, number, "indentation must be exactly two spaces per level")
        lines.append(_Line(text, number, indent, text[indent:]))
    return lines


class _Parser:
    def __init__(self, lines: list[_Line], source: str | Path):
        self.lines = lines
        self.source = source
        self.nodes = 0

    def fail(self, line: int, message: str):
        raise _error(self.source, line, message)

    def new_node(self, line: int):
        self.nodes += 1
        if self.nodes > MAX_NODES:
            self.fail(line, f"parsed node count exceeds {MAX_NODES}")

    def significant(self, pos: int) -> int:
        while pos < len(self.lines) and self.lines[pos].ignored:
            pos += 1
        return pos

    def parse(self) -> dict:
        pos = self.significant(0)
        if pos == len(self.lines):
            line = self.lines[0].number if self.lines else 1
            self.fail(line, "top-level mapping must be non-empty")
        if self.lines[pos].indent != 0:
            self.fail(self.lines[pos].number, "top-level mapping must start at column zero")
        value, end = self.parse_container(pos, 0, 1)
        end = self.significant(end)
        if end != len(self.lines):
            self.fail(self.lines[end].number, "unexpected content after top-level mapping")
        if not isinstance(value, dict):
            self.fail(self.lines[pos].number, "top-level value must be a mapping")
        return value

    def parse_container(self, pos: int, indent: int, depth: int):
        if depth > MAX_DEPTH:
            line = self.lines[pos].number if pos < len(self.lines) else 1
            self.fail(line, f"nesting depth exceeds {MAX_DEPTH}")
        pos = self.significant(pos)
        if pos >= len(self.lines):
            self.fail(self.lines[-1].number if self.lines else 1, "nested container is empty")
        line = self.lines[pos]
        if line.indent != indent:
            self.fail(line.number, f"expected indentation of {indent} spaces")
        if line.content == "-" or line.content.startswith("- "):
            return self.parse_sequence(pos, indent, depth)
        return self.parse_mapping(pos, indent, depth)

    def parse_mapping(
        self,
        pos: int,
        indent: int,
        depth: int,
        initial: tuple[str, str | None, _Line] | None = None,
    ):
        self.new_node(initial[2].number if initial else self.lines[pos].number)
        result: dict = {}
        fields = 0

        def add(key: str, raw: str | None, line: _Line, next_pos: int) -> int:
            nonlocal fields
            fields += 1
            if fields > MAX_FIELDS:
                self.fail(line.number, f"mapping has more than {MAX_FIELDS} fields")
            if key in result:
                self.fail(line.number, f"duplicate key {key!r}")
            if raw is None:
                child = self.significant(next_pos)
                if child >= len(self.lines) or self.lines[child].indent <= indent:
                    self.fail(line.number, f"bare key {key!r} has no nested container")
                if self.lines[child].indent != indent + 2:
                    self.fail(
                        self.lines[child].number,
                        f"nested value for {key!r} must be indented exactly two spaces",
                    )
                value, consumed = self.parse_container(child, indent + 2, depth + 1)
                result[key] = value
                return consumed
            if raw == ">":
                value, consumed = self.parse_folded(next_pos, indent, line)
                result[key] = value
                self.new_node(line.number)
                return consumed
            result[key] = self.parse_scalar(raw, line.number)
            self.new_node(line.number)
            return next_pos

        if initial:
            pos = add(*initial, pos)
        while True:
            pos = self.significant(pos)
            if pos >= len(self.lines):
                break
            line = self.lines[pos]
            if line.indent < indent:
                break
            if line.indent > indent:
                self.fail(line.number, f"ambiguous indentation; expected {indent} spaces")
            if line.content == "-" or line.content.startswith("- "):
                self.fail(line.number, "mapping and sequence entries cannot mix")
            match = _KEY_VALUE_RE.fullmatch(line.content)
            if not match:
                self.fail(line.number, "expected 'key: value' mapping entry")
            pos = add(match.group(1), match.group(2), line, pos + 1)
        if not result:
            line = initial[2].number if initial else self.lines[pos].number
            self.fail(line, "mapping must be non-empty")
        return result, pos

    def parse_sequence(self, pos: int, indent: int, depth: int):
        self.new_node(self.lines[pos].number)
        result: list = []
        item_kind: str | None = None
        while True:
            pos = self.significant(pos)
            if pos >= len(self.lines):
                break
            line = self.lines[pos]
            if line.indent < indent:
                break
            if line.indent > indent:
                self.fail(line.number, f"ambiguous indentation; expected {indent} spaces")
            if line.content == "-":
                self.fail(line.number, "a bare sequence marker has no value")
            if not line.content.startswith("- "):
                self.fail(line.number, "mapping and sequence entries cannot mix")
            if len(result) >= MAX_SEQUENCE_ENTRIES:
                self.fail(line.number, f"sequence has more than {MAX_SEQUENCE_ENTRIES} entries")
            raw = line.content[2:]
            mapping = _KEY_VALUE_RE.fullmatch(raw)
            kind = "mapping" if mapping else "scalar"
            if item_kind is not None and kind != item_kind:
                self.fail(line.number, "scalar and mapping sequence entries cannot mix")
            item_kind = kind
            if mapping:
                value, pos = self.parse_mapping(
                    pos + 1,
                    indent + 2,
                    depth + 1,
                    (mapping.group(1), mapping.group(2), line),
                )
                result.append(value)
            else:
                if raw.strip() == "[]":
                    self.fail(line.number, "sequences cannot contain nested sequences")
                result.append(self.parse_scalar(raw, line.number))
                self.new_node(line.number)
                pos += 1
        return result, pos

    def parse_folded(self, pos: int, parent_indent: int, owner: _Line):
        parts: list[str] = []
        while pos < len(self.lines):
            line = self.lines[pos]
            if not line.content:
                if line.indent > parent_indent or (parts and pos != len(self.lines) - 1):
                    self.fail(line.number, "blank folded-text continuation is unsupported")
                break
            if line.indent <= parent_indent:
                break
            if line.indent != parent_indent + 2:
                self.fail(
                    line.number, "folded text must be indented exactly two spaces"
                )
            if line.content.startswith("#"):
                pos += 1
                continue
            parts.append(line.content.strip())
            pos += 1
        if not parts:
            self.fail(owner.number, "folded text must have a non-empty continuation")
        value = " ".join(parts)
        self.check_scalar_size(value, owner.number)
        return value, pos

    def check_scalar_size(self, value: str, line: int):
        for char in value:
            codepoint = ord(char)
            if codepoint < 0x20 and char != "\n":
                self.fail(line, f"scalar contains unsupported C0 control 0x{codepoint:02x}")
        try:
            encoded = value.encode("utf-8")
        except UnicodeEncodeError:
            self.fail(line, "scalar contains an unsupported Unicode surrogate")
        if len(encoded) > MAX_SCALAR_BYTES:
            self.fail(line, f"scalar exceeds {MAX_SCALAR_BYTES} bytes")

    def parse_scalar(self, raw: str, line: int):
        value = raw.strip()
        if value == "":
            self.fail(line, 'empty scalar must be written as ""')
        if value == "[]":
            return []
        if value.startswith("|"):
            self.fail(line, "literal block scalar '|' is unsupported")
        if value.startswith(">"):
            self.fail(line, "folded-text chomping indicators are unsupported")
        if value.startswith("'"):
            self.fail(line, "single-quoted scalars are unsupported")
        if value.startswith('"'):
            try:
                decoded = json.loads(value)
            except json.JSONDecodeError:
                self.fail(line, "double-quoted scalar must use JSON string escaping")
            if not isinstance(decoded, str):
                self.fail(line, "quoted scalar must decode to a string")
            self.check_scalar_size(decoded, line)
            return decoded
        if value.startswith(_UNSUPPORTED_PREFIXES) or value == "<<" or value.startswith("<<:"):
            self.fail(line, "anchors, aliases, tags, merge keys, and directives are unsupported")
        if value in ("---", "..."):
            self.fail(line, "multiple documents are unsupported")
        if any(char in value for char in "{}[]"):
            self.fail(line, "flow-style collections are unsupported")
        self.check_scalar_size(value, line)
        return value


def parse_frontmatter(data: bytes, source: str | Path = "<bytes>") -> dict:
    """Parse one fenced Markdown frontmatter mapping from raw bytes."""
    if not isinstance(data, bytes):
        raise TypeError("parse_frontmatter expects bytes")
    _validate_file_bytes(data, source)
    region, first_line = _fenced_region(data, source)
    lines = _validate_region_bytes(region, source, first_line)
    _validate_utf8(data, source)
    return _Parser(lines, source).parse()


def parse_document(data: bytes, source: str | Path = "<bytes>") -> dict:
    """Parse the fence-less top-level mapping used by §21.2 plan extracts."""
    if not isinstance(data, bytes):
        raise TypeError("parse_document expects bytes")
    _validate_file_bytes(data, source)
    if len(data) > MAX_DOCUMENT_BYTES:
        raise _error(source, 1, f"frontmatter exceeds {MAX_DOCUMENT_BYTES} bytes")
    lines = _validate_region_bytes(data, source, 1)
    for line in lines:
        if line.indent == 0 and line.content in ("---", "..."):
            raise _error(source, line.number, "fences and multiple documents are unsupported")
        if line.indent == 0 and line.content.startswith("%"):
            raise _error(source, line.number, "directives are unsupported")
    return _Parser(lines, source).parse()


__all__ = [
    "FrontmatterError",
    "MAX_DEPTH",
    "MAX_DOCUMENT_BYTES",
    "MAX_FIELDS",
    "MAX_FILE_BYTES",
    "MAX_LINE_BYTES",
    "MAX_NODES",
    "MAX_SCALAR_BYTES",
    "MAX_SEQUENCE_ENTRIES",
    "parse_document",
    "parse_frontmatter",
]

File: scripts/differential/frontmatter.ts
import { readdirSync, readFileSync, statSync } from "node:fs";
import { oracleAnswer } from "./oracle.ts";

import { join } from "node:path";

import { stringifyRow } from "../src/boundary/canonical-json.ts";
import {
  FrontmatterError,
  frontmatterBody,
  MAX_DEPTH,
  MAX_DOCUMENT_BYTES,
  MAX_FIELDS,
  MAX_FILE_BYTES,
  MAX_LINE_BYTES,
  MAX_NODES,
  MAX_SCALAR_BYTES,
  MAX_SEQUENCE_ENTRIES,
  parseDocument,
  parseFrontmatter,
} from "../src/boundary/frontmatter.ts";

// Compares the ported §20.4 grammar against the reference parser on three
// axes at once: whether a document is accepted, the exact diagnostic when it
// is not, and the parsed value when it is. A grammar this closed is defined
// as much by what it refuses as by what it returns, so an accept/reject-only
// comparison would miss a parser that refuses the right documents for the
// wrong reason — and the message is what a caller sees.

interface Case {
  readonly label: string;
  readonly source: string;
  readonly bytes: Uint8Array;
}

const cases: Case[] = [];
const encoder = new TextEncoder();

function add(label: string, text: string, source = "<bytes>"): void {
  cases.push({ label, source, bytes: encoder.encode(text) });
}

function addBytes(label: string, bytes: number[], source = "<bytes>"): void {
  cases.push({ label, source, bytes: Uint8Array.from(bytes) });
}

function fenced(body: string): string {
  return `---\n${body}\n---\nbody text\n`;
}

// ---- documents the grammar accepts -----------------------------------------

add("minimal", fenced("id: a"));
add("several scalars", fenced("id: a\nkind: concept\nfield: knowledge"));
add("quoted scalar", fenced('id: "a b"\ntitle: "quoted \\"inner\\" text"'));
add("empty string scalar", fenced('id: ""'));
add("empty sequence scalar", fenced("tags: []"));
add("scalar sequence", fenced("tags:\n  - one\n  - two"));
add("mapping sequence", fenced("items:\n  - id: a\n    kind: concept\n  - id: b\n    kind: concept"));
add("nested mapping", fenced("outer:\n  inner:\n    leaf: value"));
add("folded text", fenced("note: >\n  first line\n  second line"));
add("folded then sibling", fenced("note: >\n  folded\nother: value"));
add("folded with comment", fenced("note: >\n  first\n  # not a comment marker at parse\n  second"));
add("comments and blanks", fenced("# leading\n\nid: a\n\n# trailing"));
add("unicode scalar", fenced("title: привет мир 🜂"));
add("value with colon", fenced("note: a: b"));
add("value with hash", fenced("note: a # b"));
add("key with underscore and dash", fenced("a_b-c: value"));
add("deep but legal nesting", fenced("a:\n  b:\n    c:\n      d:\n        e:\n          f:\n            g: leaf"));
add("no body after fence", "---\nid: a\n---\n");
add("body absent entirely", "---\nid: a\n---");

// ---- keys that a JavaScript object treats as more than keys ------------------
// `__proto__` satisfies the key grammar, so a document may carry it. On a plain
// `{}` it is an accessor rather than a member: the field disappears and the
// values under it become the prototype every later lookup falls back to. The
// oracle's dict has no such corner, which is precisely why these are compared.

add("proto key alone", fenced("__proto__: value"));
add("proto key beside a real one", fenced("id: a\n__proto__: value"));
add("proto key carrying a mapping", fenced("__proto__:\n  sensitivity: medical"));
add("proto key nested", fenced("outer:\n  __proto__: value"));
add("proto key in a sequence item", fenced("items:\n  - __proto__: value"));
add("proto key twice", fenced("__proto__: a\n__proto__: b"));
add("proto key and a near miss", fenced("__proto__: a\n__proto__x: b\nproto: c"));
add("constructor key", fenced("constructor: value"));
add("prototype key", fenced("prototype: value"));
add("object method names as keys", fenced("toString: a\nvalueOf: b\nhasOwnProperty: c"));

// ---- the strip-set traps ----------------------------------------------------
// Python's str.strip() removes U+0085 and U+00A0 but not U+FEFF; JavaScript's
// trim() does the reverse. Each of these decides whether a scalar keeps a
// character or loses it, so they are compared rather than reasoned about.

add("scalar padded with NBSP", fenced("note: \u00a0value\u00a0"));
add("scalar padded with NEL", fenced("note: \u0085value\u0085"));
add("scalar padded with ZWNBSP", fenced("note: \ufeffvalue\ufeff"));
add("scalar padded with ideographic space", fenced("note: \u3000value\u3000"));
add("scalar padded with en quad", fenced("note: \u2000value\u2000"));
add("scalar of only NBSP", fenced("note: \u00a0"));
add("scalar of only ZWNBSP", fenced("note: \ufeff"));
add("zero-width space is not whitespace", fenced("note: \u200bvalue\u200b"));
add("folded line padded with NBSP", fenced("note: >\n  \u00a0folded\u00a0"));
add("sequence entry padded with NBSP", fenced("tags:\n  - \u00a0value\u00a0"));
add("nested-sequence probe with NBSP", fenced("tags:\n  - \u00a0[]\u00a0"));

// ---- refusals ---------------------------------------------------------------

add("no opening fence", "id: a\n");
add("opening fence with trailing space", "--- \nid: a\n---\n");
add("missing closing fence", "---\nid: a\n");
add("empty frontmatter", "---\n---\n");
add("only comments", fenced("# just a comment"));
add("top level indented", "---\n  id: a\n---\n");
add("odd indentation", fenced("outer:\n   inner: value"));
add("over-indented nested value", fenced("outer:\n    inner: value"));
add("bare key without container", fenced("outer:"));
add("bare key then sibling", fenced("outer:\nother: value"));
add("duplicate key", fenced("id: a\nid: b"));
add("duplicate key nested", fenced("outer:\n  a: 1\n  a: 2"));
add("mixed mapping and sequence", fenced("id: a\n- item"));
add("bare sequence marker", fenced("tags:\n  -"));
add("mixed sequence item kinds", fenced("tags:\n  - one\n  - id: a"));
add("mixed sequence item kinds reversed", fenced("tags:\n  - id: a\n  - one"));
add("nested sequence", fenced("tags:\n  - []"));
add("empty scalar", fenced("id:  "));
add("single quoted", fenced("id: 'a'"));
add("literal block", fenced("id: |"));
add("folded chomping", fenced("id: >-"));
add("flow mapping", fenced("id: {a: b}"));
add("flow sequence", fenced("id: [a, b]"));
add("anchor", fenced("id: &anchor"));
add("alias", fenced("id: *alias"));
add("tag", fenced("id: !!str x"));
add("directive value", fenced("id: %YAML"));
add("merge key", fenced("id: <<"));
add("merge key colon", fenced("id: <<: x"));
add("document marker scalar", fenced("id: ---"));
add("document end scalar", fenced("id: ..."));
add("bad quoted escape", fenced('id: "bad \\q escape"'));
add("quoted decoding to non-string", fenced('id: "a" extra'));
add("lone surrogate escape", fenced('id: "\\ud800"'));
add("key starting with digit", fenced("1key: value"));
add("key with space", fenced("a key: value"));
add("colon without space", fenced("id:value"));
add("too deep nesting", fenced("a:\n  b:\n    c:\n      d:\n        e:\n          f:\n            g:\n              h: leaf"));
add("blank folded continuation", fenced("note: >\n  first\n\n  second"));
add("folded with no continuation", fenced("note: >\nother: value"));
add("folded over-indented", fenced("note: >\n    first"));
add("unexpected trailing content", fenced("id: a\n\nsecond: b\n"));

// ---- byte-level refusals ----------------------------------------------------

addBytes("utf-8 BOM", [0xef, 0xbb, 0xbf, ...encoder.encode("---\nid: a\n---\n")]);
addBytes("CR in region", [...encoder.encode("---\nid: a\r\n---\n")]);
addBytes("CR in body", [...encoder.encode("---\nid: a\n---\nbody\r\n")]);
addBytes("tab in region", [...encoder.encode("---\nid:\ta\n---\n")]);
addBytes("NUL in region", [...encoder.encode("---\nid: a"), 0x00, ...encoder.encode("\n---\n")]);
addBytes("C0 control in region", [...encoder.encode("---\nid: a"), 0x01, ...encoder.encode("\n---\n")]);
addBytes("bare continuation byte in region", [...encoder.encode("---\nid: "), 0x80, ...encoder.encode("\n---\n")]);
addBytes("truncated sequence in region", [...encoder.encode("---\nid: "), 0xe2, 0x82, ...encoder.encode("\n---\n")]);
addBytes("overlong encoding in region", [...encoder.encode("---\nid: "), 0xc0, 0x80, ...encoder.encode("\n---\n")]);
addBytes("surrogate encoding in region", [...encoder.encode("---\nid: "), 0xed, 0xa0, 0x80, ...encoder.encode("\n---\n")]);
addBytes("out of range lead in region", [...encoder.encode("---\nid: "), 0xf5, 0x80, 0x80, 0x80, ...encoder.encode("\n---\n")]);
addBytes("invalid utf-8 in body", [...encoder.encode("---\nid: a\n---\nbody "), 0xff, 0x0a]);
addBytes("invalid utf-8 later in body", [...encoder.encode("---\nid: a\n---\nline one\nline two "), 0xc3, 0x28, 0x0a]);
addBytes("tab in body is allowed", [...encoder.encode("---\nid: a\n---\nbody\twith tab\n")]);

// ---- ceilings ---------------------------------------------------------------

add("line at the ceiling", fenced(`note: ${"x".repeat(4_096 - 6)}`));
add("line over the ceiling", fenced(`note: ${"x".repeat(4_096)}`));
add("scalar over the ceiling", fenced(`note: ${"x".repeat(8_193)}`));
add("too many fields", fenced(
  Array.from({ length: 65 }, (_, i) => `k${i}: v`).join("\n"),
));
add("fields at the ceiling", fenced(
  Array.from({ length: 64 }, (_, i) => `k${i}: v`).join("\n"),
));
add("too many sequence entries", fenced(
  "tags:\n" + Array.from({ length: 1_025 }, (_, i) => `  - v${i}`).join("\n"),
));
add("frontmatter over the document ceiling", fenced(
  Array.from({ length: 2_000 }, (_, i) => `k${i}: ${"v".repeat(70)}`).join("\n"),
));

// ---- source labelling -------------------------------------------------------

add("named source in diagnostic", "---\nid: a\n", "atlas/concepts/a.md");
add("named source accepted", fenced("id: a"), "atlas/concepts/a.md");

// ---- the repository's own documents -----------------------------------------
// Hand-written cases probe the edges; these prove the parser on the corpus the
// project actually has, where a subtle regression would otherwise surface only
// as a failed build much later.

function collectMarkdown(dir: string, out: string[]): void {
  let entries: string[];
  try {
    entries = readdirSync(dir);
  } catch {
    return;
  }
  for (const entry of entries) {
    const path = join(dir, entry);
    const info = statSync(path);
    if (info.isDirectory()) collectMarkdown(path, out);
    else if (entry.endsWith(".md")) out.push(path);
  }
}

const fixtures: string[] = [];
collectMarkdown("fixtures", fixtures);
fixtures.sort();
for (const path of fixtures) {
  cases.push({
    label: `fixture ${path}`,
    source: path,
    bytes: new Uint8Array(readFileSync(path)),
  });
}

// ---- comparison -------------------------------------------------------------

interface Outcome {
  ok: boolean;
  value?: string;
  error?: string;
}

interface OracleEntry {
  frontmatter: Outcome;
  document: Outcome;
  body: Outcome;
}

function toBase64(bytes: Uint8Array): string {
  return Buffer.from(bytes).toString("base64");
}

// `validate_atlas check-constants` ties the reference parser's ceilings to the
// schemas. Nothing ties this copy to either, so the values are compared here
// directly rather than only through the behaviour a few of them govern.
const CONSTANTS: Record<string, number> = {
  MAX_DOCUMENT_BYTES,
  MAX_FILE_BYTES,
  MAX_LINE_BYTES,
  MAX_SCALAR_BYTES,
  MAX_DEPTH,
  MAX_FIELDS,
  MAX_SEQUENCE_ENTRIES,
  MAX_NODES,
};

const payload = JSON.stringify({
  cases: cases.map((entry) => ({
    data: toBase64(entry.bytes),
    source: entry.source,
  })),
  constants: Object.keys(CONSTANTS),
});

const parsed = oracleAnswer("frontmatter", payload) as {
  cases: OracleEntry[];
  constants: Record<string, number>;
};
const oracle = parsed.cases;

let divergences = 0;
let compared = 0;

function run(fn: () => unknown): Outcome {
  try {
    const value = fn();
    return { ok: true, value: stringifyRow(value) };
  } catch (err) {
    if (err instanceof FrontmatterError) return { ok: false, error: err.message };
    return {
      ok: false,
      error: `UNEXPECTED ${(err as Error).constructor.name}: ${String(err)}`,
    };
  }
}

function compare(label: string, actual: Outcome, wanted: Outcome): void {
  compared += 1;
  const same = actual.ok === wanted.ok &&
    (actual.ok ? actual.value === wanted.value : actual.error === wanted.error);
  if (!same) {
    divergences += 1;
    console.error(`DIVERGENCE ${label}`);
    console.error(`  oracle: ${JSON.stringify(wanted)}`);
    console.error(`  ours:   ${JSON.stringify(actual)}`);
  }
}

for (let i = 0; i < cases.length; i += 1) {
  const entry = cases[i]!;
  const expected = oracle[i]!;
  compare(
    `parse_frontmatter ${entry.label}`,
    run(() => parseFrontmatter(entry.bytes, entry.source)),
    expected.frontmatter,
  );
  compare(
    `parse_document ${entry.label}`,
    run(() => parseDocument(entry.bytes, entry.source)),
    expected.document,
  );
  compared += 1;
  let body: string | null = null;
  try {
    body = frontmatterBody(entry.bytes);
  } catch {
    body = null;
  }
  const wantedBody = expected.body.ok ? expected.body.value : null;
  if (body !== wantedBody) {
    divergences += 1;
    console.error(`DIVERGENCE frontmatter_body ${entry.label}`);
    console.error(`  oracle: ${JSON.stringify(wantedBody)}`);
    console.error(`  ours:   ${JSON.stringify(body)}`);
  }
}

for (const [name, ours] of Object.entries(CONSTANTS)) {
  compared += 1;
  const wanted = parsed.constants[name];
  if (ours !== wanted) {
    divergences += 1;
    console.error(`DIVERGENCE constant ${name}`);
    console.error(`  oracle: ${wanted}`);
    console.error(`  ours:   ${ours}`);
  }
}

console.log(
  `differential frontmatter: ${cases.length} documents, ` +
    `${compared} comparisons, ${divergences} divergences ` +
    `(${fixtures.length} from fixtures/)`,
);
process.exit(divergences === 0 ? 0 : 1);
